feat: replace github.com/denisenkom/go-mssqldb with github.com/microsoft/go-mssqldb

powersj opened the pull request:
fixes: #12745

powersj commented:
@minwal, @AsafMah, 

I need some help ensuring this PR does not regress you or our users of the azure data explorer output. This PR updates the go-mssqldb from github.com/denisenkom/go-mssqldb to github.com/microsoft/go-mssqldb.

Most of the changes are drop in except the following commit to azure data explorer output: https://github.com/influxdata/telegraf/pull/13524/commits/00cb07ed8b867133ebb4d7f2648883d138270db4

There are two changes I am concerned about:

1. The way we login. I did not see a drop in replacement for `NewAuthorizerFromEnvironmentWithResource`. The `WithDefaultAzureCredential` does state that it looks at "The order of auth will be via environment variables, managed identity and Azure CLI." Which I think aligns with the current behavior.
2. The create table and table mappings, `kuso.NewStmt` is deprecated and replaced with `kql.New()`. I was not sure if my updates are enough given I am missing the suppress warnings statement that was previously there.

Can either of you try this out or verify that I have made the correct changes?


@soccypowa,

Since you were also interested in the replacement of the go-msssqldb dependency, can you give this a shot to verify your kerberos support works?

Thanks!

soccypowa commented:
@powersj 

I made a build of your branch and took it for a spin. For some reason though I got only NTLM auth in my test environment, realised that I get the same behaviour with my own updated "[ugly build](https://github.com/soccypowa/telegraf/tree/kerberos-230629)" go-mssqldb v1.3.0, while my [old one](https://github.com/soccypowa/telegraf/tree/mssql-fix), made in the excact same way with go-mssqldb v0.17.1-0.20221011173445-537893aca440 just works. I tried to do a fast "read up" on wether MSFT changed some behaviour or not, I mean they have changed default settings in many of their other drivers so why not this one. Have not found any such indications, but if my working schedule allows me I will look into this, in any case out of curiosity. Nice work getting the change in place though!

srebhan commented on go.mod:
```diff
@@ -484,3 +484,5 @@ require (
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
+
+replace github.com/microsoft/go-mssqldb v1.3.0 => github.com/powersj/go-mssqldb v1.3.0-fix2
```
Is there any PR we can watch to remove this replacement ASAP?

srebhan reviewed (approved):
Looks good to me. We should try to remove the `go.mod` replacement as soon as possible IMO...

srebhan commented:
@powersj giving this back to you in case you want to wait for the upstream PR to be merged...

powersj commented:
@srebhan Upstream PR is merged, thanks for catching the replace was still there. I've updated the go.mod to use the latest commit and removed the replace.
